database/migrations/sqls/20201231060447-domainEvents-up.sql

```diff
@@ -1,11 +1,24 @@
 /* Replace with your SQL commands */
 CREATE TABLE domain_event
 (
-    id uuid NOT NULL PRIMARY KEY,
+    id uuid NOT NULL,
     payload jsonb NOT NULL,
     status varchar NOT NULL,
     created_at timestamptz NOT NULL,
-    updated_at timestamptz NOT NULL
-);
+    updated_at timestamptz NOT NULL,
+    CONSTRAINT domain_event_pkey PRIMARY KEY (id, status, created_at)
+) PARTITION BY LIST (status);
+
 CREATE INDEX event_status_idx ON domain_event (status);
-CREATE INDEX event_type_idx ON domain_event USING GIN (payload jsonb_path_ops);
+CREATE INDEX event_pyld_idx ON domain_event USING GIN (payload jsonb_path_ops);
+

```

Comment:
We want to partition by list based on domain_event statuses and then further by range based on creation date only for those statuses that are terminal.  For e.g. `raised` and `received` statuses indicates an undispatched or unhandled events which are initial states and system would process them and move it to terminal state `sent` and `handled`. This terminal state would accumulate almost all of the data and hence we want to partition it by range based on creation date. We could then remove the paritions that are more than a year old for maintenance.

Reply:
Great, I completely agree.